libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -183,11 +185,39 @@ size_t ContractCompiler::packIntoContractCreator(ContractDefinition const& _cont
 	m_context << deployRoutine;
 
 	solAssert(m_context.runtimeSub() != size_t(-1), "Runtime sub not registered");
+
+	u256 codeMemoryStart = (u256(CompilerUtils::generalPurposeMemoryStart) + m_context.reservedMemory());
+
 	m_context.pushSubroutineSize(m_context.runtimeSub());
 	m_context << Instruction::DUP1;
 	m_context.pushSubroutineOffset(m_context.runtimeSub());
-	m_context << u256(0) << Instruction::CODECOPY;
-	m_context << u256(0) << Instruction::RETURN;
+	m_context << u256(codeMemoryStart) << Instruction::CODECOPY;
+	ContractType contractType(_contract);
+	auto const& immutables = contractType.immutableVariables();
+	for (auto const& [immutableName, occurrences]:
+		// TODO: is it ok to access the assembled runtime object here like this?
```

Comment:
Please check that the optimizer has run

Reply:
Yep, the gas cost test failures right now pretty much confirm that this is a problem. I'll try to reorder stuff.